consensus/src/consensus_observer/observer/execution_pool.rs

```diff
@@ -62,3 +69,545 @@ impl ObservedOrderedBlock {
         }
     }
 }
+
+/// Returns all ordered blocks for the given ordered block with window.
+/// This requires traversing backward via the parent links to identify
+/// and fetch the blocks from the pending block store. The blocks are
+/// returned in chronological order, and if any block is missing, this
+/// will return None.
+pub fn get_all_blocks_for_window(
+    pending_block_store: Arc<Mutex<PendingBlockStore>>,
+    ordered_block: &OrderedBlock,
+    window_size: u64,
+) -> Option<Vec<OrderedBlock>> {
+    // If the window size is 0, something is wrong. Log the error and return.
+    if window_size == 0 {
+        error!(
+            LogSchema::new(LogEntry::ConsensusObserver).message(&format!(
+                "Execution pool window size is 0 for ordered block with window: {:?}",
+                ordered_block.proof_block_info()
+            ))
+        );
+        return None;
+    }
+
+    // TODO: this doesn't support ordered blocks with multiple inner blocks.
+    // But, this should be fine, as we'd fallback to state sync anyway.
+
+    // Identify the window boundary (i.e., the lowest round that falls inside the window)
+    let ordered_block_round = ordered_block.first_block().round();
+    let window_boundary_round = ordered_block_round
+        .saturating_add(1)
+        .saturating_sub(window_size);
+
+    // Add the current block to the window of all ordered blocks
+    let mut all_ordered_blocks = vec![ordered_block.clone()];
+
+    // Collect all ordered blocks for the window
+    let mut current_block = ordered_block.clone();
+    let mut remaining_window_size = window_size.saturating_sub(1);
+    while remaining_window_size > 0 {
```

Comment:
The exact window definition changed since we last chatted.

see: block_tree: `get_ordered_block_window()`

It would be great to share code, but if that's not possible the logic still needs to be the same. Perhaps the unit tests can construct a block tree, an equivalent pending block store, and check the equivalence?

Reply:
I've pulled the code out and reused it. Hopefully it's clear enough.